From 30ab9445d4fa3dfe83d066706a0d23a54f47d7e7 Mon Sep 17 00:00:00 2001
Message-Id: <30ab9445d4fa3dfe83d066706a0d23a54f47d7e7.1414143271.git.chang-joon.lee@intel.com>
In-Reply-To: <f393e797ccaaf66a25ee3d7c4e6136083a6effeb.1414143271.git.chang-joon.lee@intel.com>
References: <f393e797ccaaf66a25ee3d7c4e6136083a6effeb.1414143271.git.chang-joon.lee@intel.com>
From: Yogesh Mohan Marimuthu <yogesh.mohan.marimuthu@intel.com>
Date: Wed, 15 Oct 2014 18:53:29 +0530
Subject: [PATCH 06/26] REVERTME [VPG]: hardcode 180deg wa for DSI port C too

Till CHT A2 stepping there is a hardware bug that data and
clock are out of phase at 180deg. Hardcoding the delay registers
at 900Mhz DSI PLL frequency to fix the 180deg shift.

REVERTME: This patch can be reverted once everyone in Intel has
stopped using CHT SoC steppings older than A3.

For: GMINL-1532
Change-Id: I5bbf6d74a637671c840b190d15b38b9d560b45a9
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
Signed-off-by: Yogesh Mohan Marimuthu <yogesh.mohan.marimuthu@intel.com>
---
 drivers/gpu/drm/i915/i915_reg.h  |    3 +++
 drivers/gpu/drm/i915/intel_dsi.c |   11 +++++++++--
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index e7eafc4..8cd5b0c 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -6385,6 +6385,9 @@ enum punit_power_well {
 #define  TEARING_EFFECT_DELAY_SHIFT			0
 #define  TEARING_EFFECT_DELAY_MASK			(0xffff << 0)
 
+#define DELAY_180_PHASE_SHIFT_MIPIA			0x682C0000
+#define DELAY_180_PHASE_SHIFT_MIPIC			0xF0E0
+
 /* XXX: all bits reserved */
 #define _MIPIA_AUTOPWG				(VLV_DISPLAY_BASE + 0x611a0)
 
diff --git a/drivers/gpu/drm/i915/intel_dsi.c b/drivers/gpu/drm/i915/intel_dsi.c
index 54d39a6..204c0d3 100644
--- a/drivers/gpu/drm/i915/intel_dsi.c
+++ b/drivers/gpu/drm/i915/intel_dsi.c
@@ -110,8 +110,17 @@ static void intel_dsi_device_ready(struct intel_encoder *encoder)
 	DRM_DEBUG_KMS("\n");
 
 	val = I915_READ(MIPI_PORT_CTRL(0));
+
+	/*
+	 * Putting delay of 180 phase shift for MIPI Port A & Port C
+	 * Values given by SV team
+	 */
+	if (IS_CHERRYVIEW(dev_priv->dev) && STEP_TO(STEP_A2))
+		val |= pipe ? DELAY_180_PHASE_SHIFT_MIPIC :
+				DELAY_180_PHASE_SHIFT_MIPIA;
 	I915_WRITE(MIPI_PORT_CTRL(0), val | LP_OUTPUT_HOLD);
 	usleep_range(1000, 1500);
+
 	I915_WRITE(MIPI_DEVICE_READY(pipe), DEVICE_READY | ULPS_STATE_EXIT);
 	usleep_range(2000, 2500);
 	I915_WRITE(MIPI_DEVICE_READY(pipe), DEVICE_READY);
@@ -148,8 +157,6 @@ static void intel_dsi_enable(struct intel_encoder *encoder)
 		/* assert ip_tg_enable signal */
 		temp = I915_READ(MIPI_PORT_CTRL(pipe)) & ~LANE_CONFIGURATION_MASK;
 		temp = temp | intel_dsi->port_bits;
-		if (IS_CHERRYVIEW(dev))
-			temp |= 0xe82d0000;
 		I915_WRITE(MIPI_PORT_CTRL(pipe), temp | DPI_ENABLE);
 		POSTING_READ(MIPI_PORT_CTRL(pipe));
 	}
-- 
1.7.9.5

